### PR TITLE
ID Datastructure rule of 6

### PR DESCRIPTION
**Pull request by timosachsenberg**
## Description

add move const. and assignment to ID datastructure

## Checklist
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

File: src/tests/class_tests/openms/source/IdentificationData_test.cpp [hunk 1]
```diff
@@ -60,6 +62,12 @@ START_SECTION((IdentificationData()))
   TEST_NOT_EQUAL(ptr, null);
 END_SECTION
 
+START_SECTION((movable))
+  TEST_EQUAL(std::is_nothrow_move_constructible<IdentificationData>::value, true);
+  TEST_EQUAL(std::is_nothrow_move_assignable<IdentificationData>::value, true);
```
Comment (cbielow) [inline src/tests/class_tests/openms/source/IdentificationData_test.cpp:67]: nice to have, thanks!

Comment (cbielow) [inline src/tests/class_tests/openms/source/IdentificationData_test.cpp:67]:
```suggestion
  TEST_TRUE(std::is_nothrow_move_constructible_v<IdentificationData>);
  TEST_TRUE(std::is_nothrow_move_assignable_v<IdentificationData>);  
```

**Review by cbielow — COMMENTED** [review]
looks like it can be simplified?!

File: src/openms/source/METADATA/ID/IdentificationData.cpp [hunk 2]
```diff
@@ -1244,6 +1244,56 @@ namespace OpenMS
...
+
+    // look-up tables:
+    observation_lookup_ = std::move(other.observation_lookup_);
+    parent_lookup_ = std::move(other.parent_lookup_);
+    identified_peptide_lookup_ = std::move(other.identified_peptide_lookup_);
+    identified_compound_lookup_ = std::move(other.identified_compound_lookup_);
+    identified_oligo_lookup_ = std::move(other.identified_oligo_lookup_);
+    observation_match_lookup_ = std::move(other.observation_match_lookup_);
+
+    return *this;
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/METADATA/ID/IdentificationData.cpp:1295]: it can't be noexcept defaulted

[the remark this answers is not in the document]
Comment (cbielow) [reply src/openms/source/METADATA/ID/IdentificationData.cpp:1295]: ahh interesting. Due to boost?

[the remark this answers is not in the document]
Comment (jpfeuffer) [reply src/openms/source/METADATA/ID/IdentificationData.cpp:1295]: Maybe VS2019? I mean you could try and see in CI.

[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/METADATA/ID/IdentificationData.cpp:1295]: yeah pretty sure VS2019. let's try

**Comment by timosachsenberg** [comment]
except for 
openms/PR/bldtst/compiler=appleclang-11.0.0,os_label=catalina it looks good.
```
<BuildLogLine>5</BuildLogLine>
<Text>/Users/builder/jenkins/ws/openms/PR/bldtst/87227eb0/source/src/openms/include/OpenMS/METADATA/ID/IdentificationData.h:254:5: error: exception specification of explicitly defaulted move constructor does not match the calculated one</Text>
<SourceFile>/Users/builder/jenkins/ws/openms/PR/bldtst/87227eb0/source/src/openms/include/OpenMS/METADATA/ID/IdentificationData.h</SourceFile>
<SourceLineNumber>254</SourceLineNumber>
<PreContext>[ 25%] Building CXX object CMakeFiles/my_custom_lib.dir/ExampleLibraryFile.cpp.o In file included from /.../external/ExampleLibraryFile.cpp:37: In file included from /Users/builder/jenkins/ws/openms/PR/bldtst/87227eb0/source/src/openms/include/OpenMS/KERNEL/Feature.h:38: In file included from /Users/builder/jenkins/ws/openms/PR/bldtst/87227eb0/source/src/openms/include/OpenMS/KERNEL/BaseFeature.h:39: </PreContext>
<PostContext> IdentificationData(IdentificationData&& other) noexcept = default; ^ </PostContext>
<RepeatCount>0</RepeatCount>
</Error>
<Error>
<BuildLogLine>8</BuildLogLine>
<Text>/Users/builder/jenkins/ws/openms/PR/bldtst/87227eb0/source/src/openms/include/OpenMS/METADATA/ID/IdentificationData.h:259:25: error: exception specification of explicitly defaulted move assignment operator does not match the calculated one</Text>
<SourceFile>/Users/builder/jenkins/ws/openms/PR/bldtst/87227eb0/source/src/openms/include/OpenMS/METADATA/ID/IdentificationData.h</SourceFile>
<SourceLineNumber>259</SourceLineNumber>
<PreContext/>
<PostContext> IdentificationData& operator=(IdentificationData&& other) noexcept = default; ^ 2 errors generated. [ 50%] Linking CXX static library libmy_custom_lib.a </PostContext>
<RepeatCount>0</RepeatCount>
</Error>
<Error>
```

**Comment by timosachsenberg** [comment]
/revert [5d1ce54](https://github.com/OpenMS/OpenMS/pull/6736/commits/5d1ce54474cc5f545a8beef9f7cf440a4a0d8659)

**Pull request merged**